lib/mongodb.js

```diff
@@ -768,45 +768,100 @@ function convertToMeters(distance, unit) {
-    var s = near.near.split(',');
-    coordinates.lng = Number(s[0]);
-    coordinates.lat = Number(s[1]);
-  } else {
-    coordinates = near.near;
+function buildNearFilter(query, params) {
+  if (!Array.isArray(params)) {
+    params = [params];
   }
 
-  query.where[near.key] = {
-    near: {
-      $geometry: {
-        coordinates: [coordinates.lng, coordinates.lat],
-        type: 'Point',
+  params.forEach(function(near) {
+    var coordinates = {};
+
+    if (typeof near.near === 'string') {
+      var s = near.near.split(',');
+      coordinates.lng = Number(s[0]);
+      coordinates.lat = Number(s[1]);
+    } else {
+      coordinates = near.near;
+    }
+
+    var props = ['maxDistance', 'minDistance'];
+    //use mongodb default unit 'meters' rather than 'miles'
+    var unit = near.unit || 'meters';
+
+    var queryValue = {
+      near: {
+        $geometry: {
+          coordinates: [coordinates.lng, coordinates.lat],
+          type: 'Point',
+        },
       },
-    },
-  };
+    };
+
+    props.forEach(function(p) {
+      if (near[p]) {
+        queryValue['near']['$' + p] = convertToMeters(near[p], unit);
+      }
+    });
+
+    var property;
 
-  var props = ['maxDistance', 'minDistance'];
-  //use mongodb default unit 'meters' rather than 'miles'
-  var unit = near.unit || 'meters';
-  props.forEach(function(p) {
-    if (near[p]) {
-      query.where[near.key]['near']['$' + p] = convertToMeters(near[p], unit);
+    if (near.key) {
+      if (Array.isArray(near.key)) {
+        property = query.where;
+
+        for (var i = 0; i < near.key.length; i++) {
+          var subKey = near.key[i];
+          if (near.key.hasOwnProperty(i + 1)) {
+            if (!property.hasOwnProperty(subKey)) {
+              property[subKey] = Number.isInteger(near.key[i + 1]) ? [] : {};
+            }
+
+            property = property[subKey];
+          }
+        }
+
+        property[near.key[i - 1]] = queryValue;
+      } else {
+        property = query.where[near.key] = queryValue;
+      }
     }
   });
 };
 
 function hasNearFilter(where) {
   if (!where) return false;
+  // TODO: Optimize to return once a `near` key is found
+  // instead of searching through everything
+
+  var found = false;
+
+  searchForNear(where);
 
-  for (var k in where) {
-    if (where[k] && typeof where[k] === 'object' && where[k].near) {
-      return true;
+  function searchForNear(node) {
+    if (!node) {
+      return;
+    }
+    if (Array.isArray(node)) {
+      node.forEach(function(nodeProperty, key) {
+        if (nodeProperty && nodeProperty.near) {
+          found = true;
+        } else {
+          searchForNear(nodeProperty);
+        }
+      });
+    } else if (typeof node === 'object') {
+      Object.keys(node).forEach(function(key) {
+        var nodeProperty = node[key];
+
+        if (nodeProperty && nodeProperty.near) {
+          found = true;
+        } else {
+          searchForNear(nodeProperty);
+        }
```

Comment:
we should remove this duplication:

```

function nearFound(prop) {
  return prop && prop.near;
}
```

then in both cases above, you can just pass in:

```
Object.keys(node).forEach(function(key) {
  var prop = node[key];
  isFound = found(prop);
  if (!isFound)
    searchForNear(prop);
});
```

And reuse found() for the array check too